[Windows] Allow exceptions through to the OS

staticfloat opened the pull request:
In order to properly generate minidumps during actual events, it is necessary to disable the call to `SetErrorMode()` during libuv initialization and to remove the job flag
`JOB_OBJECT_LIMIT_DIE_ON_UNHANDLED_EXCEPTION`, both of which interfered in the normal handling of exceptions.  Finally, it was determined prudent to include `CREATE_DEFAULT_ERROR_MODE` when spawning new processes, to ensure that whatever error handling mode was set in the current process is not inherited by subprocesses.

tkelman commented:
automake takes about 5 minutes to compile a local version of, would make the diff a lot less noisy

kpamnany commented:
Not really much less noisy, but okay.

tkelman commented:
Looks nicer to me, thanks. I think we could make a test branch of Julia and run appveyor on this if it would help.

kpamnany commented:
If that's the best way to get this tested on Windows, then sure. In any case, I'll work on building Julia in Windows per the instructions in ``README.windows``.

tkelman commented:
This looks okay for me in local builds, both win32 and win64, so if you and/or upstream libuv are happy with it I think we could just merge (unless anyone has any substantive feedback). I'm guessing that stock Julia's tests won't really execute the new features here?

kpamnany commented:
Thanks for checking! No word on the PR in upstream libuv, though the API is based on input from one of the libuv devs.

This is for use by the ``threading`` branch, to abstract away platform specifics. There is no specific test case for thread affinitization as wrong affinities will only impact performance, not correctness.

FYI, we'll need to add something for atomics as well, soon.

ArchRobison commented on src/uv-common.c:
```diff
@@ -295,6 +295,111 @@ int uv_thread_create(uv_thread_t *tid, void (*entry)(void *arg), void *arg) {
+  SetThreadAffinityMask(*tid, tam);
+
+  for (i = 0;  i < UV_CPU_SETSIZE;  i++)
+    cpumask[i] = (tam & (1 << i));
```
If i>=8, then this line will unconditionally set cpumask[i] to 0.  Did you mean cpumask[i] = (tam>>i & 1)?

kpamnany replied:
I meant ``cpumask[i] = (tam & (1 << i)) ? 1 : 0`` but your way is better. Thank you.

ArchRobison commented on src/uv-common.c:
```diff
@@ -295,6 +295,111 @@ int uv_thread_create(uv_thread_t *tid, void (*entry)(void *arg), void *arg) {
+
+  return 0;
+#else
+  int i, r;
```
I recommend moving these declarations to their points of initialization.  That way, their live range is clearer to maintainers.  Of course doing this for i requires C99.  I don't know if this package uses C99 anyway or not.

tkelman replied:
libuv generally wants things to be compile-able by oldish msvc versions that aren't c99 compliant, though for parts that are under this branch of `#ifdef _WIN32` they might be more lenient?

kpamnany replied:
From https://github.com/libuv/libuv/blob/master/CONTRIBUTING.md:

* Use C89-compliant syntax. In other words, variables can only be declared at the top of a scope (function, if/for/while-block).

ArchRobison replied:
Thanks for the citation.  Sigh.

ArchRobison commented:
Where's the documentation for these new entry points?

kpamnany commented:
Where should I document them? Upstream libuv has ``docs/`` and I will update the PR there to add these calls, but Julia's libuv appears to have been forked before documentation was added.

tkelman commented:
I'd almost say inline comments, or in this thread, for our fork. Dunno.

And just to be clear, by "This looks okay for me in local builds, both win32 and win64" I meant Julia built with this branch passes its tests. The libuv tests won't build on our fork, but this is a pre-existing problem unrelated to this PR. And I haven't tested with MSVC.

kpamnany commented:
Added the documentation I'd written in the upstream libuv PR to the top of this thread.

tkelman commented:
@vtjnash ok to merge this?

staticfloat committed:
[Windows] Allow exceptions through to the OS

In order to properly generate minidumps during actual events, it is
necessary to disable the call to `SetErrorMode()` during libuv
initialization and to remove the job flag
`JOB_OBJECT_LIMIT_DIE_ON_UNHANDLED_EXCEPTION`, both of which interfered
in the normal handling of exceptions.  Finally, it was determined
prudent to include `CREATE_DEFAULT_ERROR_MODE` when spawning new
processes, to ensure that whatever error handling mode was set in the
current process is not inherited by subprocesses.

vtjnash reviewed (approved):
Could you submit this to libuv against the master branch there?

staticfloat commented:
> Could you submit this to libuv against the master branch there?

https://github.com/libuv/libuv/pull/3838
